opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -171,11 +171,23 @@ namespace Opm
             perfRateBrine_.resize(nperf, 0.0);
             // intialize wells that have been there before
             // order may change so the mapping is based on the well name
+
+            for (int w = 0; w < nw; ++w) {
+                const Well& well = wells_ecl[w];
+                if (well.getStatus() == Well::Status::SHUT)
+                    this->shutWell(w);
+                else
+                    this->openWell(w);
+            }
+
             if (prevState && !prevState->wellMap().empty()) {
                 connpos = 0;
                 auto end = prevState->wellMap().end();
                 for (int w = 0; w < nw; ++w) {
                     const Well& well = wells_ecl[w];
+                    if (well.getStatus() == Well::Status::SHUT)
```

Comment:
This is not sufficient. You must also add:
```
                        if (!prevState->open_for_output_[oldIndex]) {
                            // Well was shut in previous state, do not use its values.
                            continue;
                        }
```
on line 196 (just after defining oldIndex).

Reply:
I think I found the last bug preventing this from exact reproduction of current master. Ironically, it was introduced by the change suggested above (and incorporated below): the `connpos` variable must be incremented for every well, but this will happen when we `continue`. (How I wish now I had persevered and merged my well state refactoring many months ago...). The solution I implemented in testing is:
```
@@ -181,7 +181,6 @@ namespace Opm
             }
 
             if (prevState && !prevState->wellMap().empty()) {
-                connpos = 0;
                 auto end = prevState->wellMap().end();
                 for (int w = 0; w < nw; ++w) {
                     const Well& well = wells_ecl[w];
@@ -232,6 +231,9 @@ namespace Opm
                         // perfPhaseRates
                         const int oldPerf_idx_beg = (*it).second[ 1 ];
                         const int num_perf_old_well = (*it).second[ 2 ];
+                        const auto new_it = wellMap().find(well.name());
+                        assert(new_it != wellMap().end());
+                        connpos = (*new_it).second[ 1 ];
                         // copy perforation rates when the number of perforations is equal,
                         // otherwise initialize perfphaserates to well rates divided by the number of perforations.
                         if( num_perf_old_well == num_perf_this_well )
@@ -294,9 +296,6 @@ namespace Opm
                     if (!has_thp) {
                         thp()[w] = 0.0;
                     }
-
-                    // Increment connection position offset.
-                    connpos += num_perf_this_well;
                 }
             }
```

Another way would be to change `if (it != end)` by adding another clause: `if ((it != end) && (prevState->open_for_output_[it->second[0]])`. That is a smaller change, but needs a comment to explain...